sdk/azidentity/chained_token_credential.go

```diff
@@ -50,35 +54,72 @@ func NewChainedTokenCredential(sources []azcore.TokenCredential, options *Chaine
 	if options == nil {
 		options = &ChainedTokenCredentialOptions{}
 	}
-	return &ChainedTokenCredential{sources: cp, name: "ChainedTokenCredential", retrySources: options.RetrySources}, nil
+	mut := sync.RWMutex{}
+	return &ChainedTokenCredential{
+		cond:         sync.NewCond(&mut),
+		mut:          &mut,
+		sources:      cp,
+		name:         "ChainedTokenCredential",
+		retrySources: options.RetrySources,
+	}, nil
 }
 
 // GetToken calls GetToken on the chained credentials in turn, stopping when one returns a token. This method is called automatically by Azure SDK clients.
 // ctx: Context controlling the request lifetime.
 // opts: Options for the token request, in particular the desired scope of the access token.
 func (c *ChainedTokenCredential) GetToken(ctx context.Context, opts policy.TokenRequestOptions) (*azcore.AccessToken, error) {
-	if c.successfulCredential != nil && !c.retrySources {
-		return c.successfulCredential.GetToken(ctx, opts)
+	if !c.retrySources {
+		// ensure only one goroutine at a time iterates the sources and perhaps sets c.successfulCredential
+		for {
+			c.mut.RLock()
+			mustIterate := c.successfulCredential == nil
+			c.mut.RUnlock()
+			if !mustIterate {
+				return c.successfulCredential.GetToken(ctx, opts)
+			}
+			c.cond.L.Lock()
+			defer c.cond.L.Unlock()
+			if !c.iterating {
+				c.iterating = true
+				break
+			}
+			c.cond.Wait()
+		}
 	}
 
+	var err error
 	var errs []error
+	var token *azcore.AccessToken
 	for _, cred := range c.sources {
-		token, err := cred.GetToken(ctx, opts)
+		token, err = cred.GetToken(ctx, opts)
 		if err == nil {
 			log.Writef(EventAuthentication, "%s authenticated with %s", c.name, extractCredentialName(cred))
-			c.successfulCredential = cred
-			return token, nil
+			if c.iterating {
```

Comment:
I think you can move `if c.iterating` to wrap around the `for` loop, then you also don't need to check it twice (similar to how we do it in expiring resource).

Reply:
I ended up with this because I don't always want to synchronize the loop. The credential may be configured to iterate its sources every time, in which case there's no reason to access `c.successfulCredential` and `c.iterating` is false.